### ISSUE TITLE
Propagate env vars into Renviron.site?

### ISSUE THREAD

**Issue by cboettig** [jupyterhub/jupyter-rsession-proxy#145]
### Bug description

Annoyingly, RStudio (though not R itself) decides to ignore global system environmental variables and only recognizes those environmental variables declared in an Renviron file (i.e. either $R_HOME/etc/Renviron.site, for all users, or a .Renviron in the user's home directory).  For instance, the client ID required by the awesome `gh-scoped-creds` python module would typically be passed in this way.  

In the rocker project, we propagate most environmental variables into R_HOME before bringing up the rserver by using the s9 init system, https://github.com/rocker-org/rocker-versioned2/blob/master/scripts/init_set_env.sh , which obviously isn't used in a jupyterhub + jupyter-rsession-proxy setup.  Would it be possible to have the jupyter-rsession-proxy do something similiar?

**Comment by cboettig** [comment]
Thanks, makes sense, I think I followed most of this.  

>  If we change a file, it'd need to be per-user and we'd have to unset the changes somehow afterwards.

I was wondering if it would be possible for juypter-ression-proxy to echo-append the env vars into the user's home dir, `~/.Renviron` ?  This wouldn't conflict with the Renviron.site coming from the docker image, and would automatically be applied per-user for Rstudio.  Not sure about the unsetting it part.  

and yeah, it's quite annoying RStudio makes us do this.

**Comment by benz0li** [comment]
With an image based on the [Jupyter Docker Stacks](https://github.com/jupyter/docker-stacks), this could be done with a [*Startup Hook*](https://jupyter-docker-stacks.readthedocs.io/en/latest/using/common.html#startup-hooks).

E.g. a `rstudio.sh` script containing

```bash
echo "LANG=$LANG" >>"$(R RHOME)/etc/Renviron.site"
echo "TZ=$TZ" >>"$(R RHOME)/etc/Renviron.site"
...
```

This would also require a change of ownership and permission of `$(R RHOME)/etc/*.site` when building the image:

```bash
chown :"$NB_GID" "$(R RHOME)/etc" "$(R RHOME)/etc/"*.site
chmod g+w "$(R RHOME)/etc" "$(R RHOME)/etc/"*.site
```

**Comment by benz0li** [comment]
@ryanlovett Is there something like *Startup Hook*s for binder, too?

**Comment by manics** [comment]
There are [`postBuild`](https://repo2docker.readthedocs.io/en/latest/config_files.html#postbuild-run-code-after-installing-the-environment) and [`start`](https://repo2docker.readthedocs.io/en/latest/config_files.html#start-run-code-before-the-user-sessions-starts) files.

**Comment by yuvipanda** [comment]
I think RStudio does this primarily to protect itself from weird env vars in people's desktops, and that causes issues when running serverside.

`jupyter-rsession-proxy` is a good place to do this!

I think the primary thing to be determined is *which file* to modify. Ideally, this should be:

1. Per-user, and not systemwide. This accounts for both permission issues, as well as the issues of running outside containerized environments.
2. At least inside containerized environments, does not persist past restarts commonly. This avoids issues possible issues with staleness.

@cboettig putting anything under `$HOME` helps with (1) but not (2). Does `Renviron` allow us to include other files? That way, another solution would be to add a line under $HOME that includes a file from somewhere else (like `/tmp` or wherever).

**Comment by yuvipanda** [comment]
Another important use case for this is with AWS credentials (or GCP credentials). These are dynamically set at runtime on the pod, and need to be propagated for automatic detection of credentials when accessing APIs to work. So while we can set gh-scoped-creds at image build time, we can't do that for these.

**Comment by cmd-ntrf** [comment]
Side note: Compute Canada / Digital Research Alliance of Canada solution for this problem was to patch RStudio. 

Patch is available here: https://github.com/ComputeCanada/easybuild-easyconfigs-installed-avx2/blob/main/2023/RStudio-Server/rstudio-1.2.1335.patch. It is quite simple, but I have never tried to have it merged upstream.

**Comment by ryanlovett** [comment]
@yuvipanda What we ended up doing was [modifying Rprofile](https://github.com/berkeley-dsep-infra/datahub/pull/5554/files) to support sourcing files in an Rprofile.d directory, and then using `extraFiles` to [set the env vars](https://github.com/berkeley-dsep-infra/datahub/pull/5558/files) there. I imagine one could add a script in Rprofile.d to source something from HOME. This is all in the user environment and outside of jupyter-rsession-proxy however.

**Comment by yuvipanda** [comment]
@ryanlovett oooh, that's interesting. Is there a per-user Rprofile? If there is, perhaps we can dynamically modify that in rsession-proxy? 

1. Modify Rprofile to load env vars from a specific location (if it exists)
2. have rsession-proxy dump out env vars in this specific location

Alternatively, we could start with putting this kinda code in Rprofile for just the rocker/binder image, where it can simply read from `/proc/1/environ`. Thoughts on that, @cboettig?

**Comment by ryanlovett** [comment]
> Is there a per-user Rprofile? If there is, perhaps we can dynamically modify that in rsession-proxy?

Yes, [~/.Rprofile](https://support.posit.co/hc/en-us/articles/360047157094-Managing-R-with-Rprofile-Renviron-Rprofile-site-Renviron-site-rsession-conf-and-repos-conf) (and also ~/.Renviron). The modification process would have to be idempotent. Would you also want to leave the vars behind? Or assume whatever file is included gets overwritten each session?

**Comment by yuvipanda** [comment]
@ryanlovett yes, we'd have the code to be idempotent - reasonably doable I'd think (`conda init` does something like that for example). 

The vars should be put on something like `/tmp`, so that gets cleaned up as appropriate. So we assume it gets overwritten each session.

**Comment by cboettig** [comment]
I think having something like:

```r
tmp <- tempfile()
writeLines(readBin("/proc/1/environ", "character", n = 500), tmp)
readRenviron(tmp)
```

in `${R_HOME}/etc/Rprofile.site` would populate the env var list when R session starts up for any user?  Or if this is handled by jupyter-session-proxy, presumably it would append this to user-specific  `$HOME/.Rprofile` ?

**Comment by benz0li** [comment]
Why not use a [Startup Hook](https://jupyter-docker-stacks.readthedocs.io/en/latest/using/common.html#startup-hooks)? e.g.

```bash
exclude_vars="HOME LD_LIBRARY_PATH OLDPWD PATH PWD RSTUDIO_VERSION SHLVL"
for var in $(compgen -e); do
  [[ ! $exclude_vars =~ $var ]] && echo "$var=${!var}" \
    >> "$(R RHOME)/etc/Renviron.site"
done
```

**Comment by yuvipanda** [comment]
@benz0li that is specific to jupyter/docker-stacks. I'd like a more general solution here.

my ideal solution is to get https://github.com/jupyterhub/jupyter-rsession-proxy/issues/145#issuecomment-2079798680 upstreamed so rstudio will also act like most other applications :D But I don't think that's going to happen.

**Comment by eeholmes** [comment]
@cboettig I think this odd Codespace devcontainer behavior is related to this.

The following devcontainer.json spins up a devcontainer with JupyterLab and RStudio.
```
{
  "name": "test",
  "workspaceFolder": "/home/jovyan",
  "image": "ghcr.io/nmfs-opensci/container-images/py-rocket-base:latest",
  "forwardPorts": [ 8889 ],
  "portsAttributes": { "8889": { "label": "Jupyter Lab",  }  },
  "postCreateCommand": "jupyter lab --ip=0.0.0.0 --port=8889 --allow-root --no-browser --NotebookApp.token='' --NotebookApp.password=''"
}
```

In the codespace from a terminal:
${PWD} is \home\jovyan

In JupyterLab from a terminal:
${PWD} is \home\jovyan

In RStudio from the terminal tab
${PWD} is \workspaces\<reponame>

In RStudio in the file panel
what is shown is \home\jovyan which is ${HOME}

If we change the workspaceFolder to something different than HOME

```
  "workspaceFolder": "/home/jovyan/codespace",
```
In the codespace from a terminal:
${PWD} is /home/jovyan/codespace

In JupyterLab from a terminal:
${PWD} is /home/jovyan/codespace

In RStudio from the terminal tab
${PWD} is \workspaces\<reponame>

In RStudio in the file panel
what is shown is \home\jovyan which is ${HOME}

**Comment by eeholmes** [comment]
Sadly,
```
echo "PWD=/home/jovyan" >> ~/.Renviron
```
seems to be ignored. Setting other envs works fine but PWD is ignored. I restarted R and restarted the terminal tab. Even setting `PWD=~` only works for that terminal. As soon as I start a new one, it goes back to the `\workspaces\<reponame>` PWD.

Also
```
echo PWD=\home\jovyan >> ~/.bashrc
```
didn't help (unless I typed `bash`).

**Comment by benz0li** [comment]
@eeholmes Set `--notebook-dir=/home/jovyan` in the `postCreateCommand`.

Cross reference: https://github.com/b-data/data-science-devcontainers#usage

**Comment by benz0li** [comment]
ℹ️ Opening your codespace in JupyterLab according to the [GitHub Docs](https://docs.github.com/en/codespaces/developing-in-codespaces/getting-started-with-github-codespaces-for-machine-learning#opening-your-codespace-in-jupyterlab) sets the default path to `/workspaces/<repository-name>` that you can not escape.

**Comment by eeholmes** [comment]
@benz0li Sadly that has no effect on the terminal being opened by `usr/bin/env bash -l` in RStudio. The PWD is fine in JupyterLab. The issue is in the terminal opened by RStudio that is being opened via the launcher.
<img width="465" alt="image" src="https://github.com/jupyterhub/jupyter-rsession-proxy/assets/2545978/31e00b55-b64a-42f5-b69c-117edcec40be">

The devcontainer.json file is working close to what I want now.
https://github.com/nmfs-opensci/container-images/blob/main/.devcontainer/test/devcontainer.json
I think I can fix the PWD issue with
```
echo -e PWD=/home/jovyan\ncd $PWD >> ~/.bash_login
```
in the postCreate command.
